core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -0,0 +1,306 @@
+                  val transactionalId: String = txnKey.key
+                  if (!record.hasValue) {
+                    loadedPids.remove(transactionalId)
+                    removedPids.add(transactionalId)
+                  } else {
+                    val txnMetadata = TransactionLog.readMessageValue(record.value)
+                    loadedPids.put(transactionalId, txnMetadata)
+                    removedPids.remove(transactionalId)
+                  }
+
+                case unknownKey =>
+                  throw new IllegalStateException(s"Unexpected message key $unknownKey while loading offsets and group metadata")
+              }
+
+              currOffset = entry.nextOffset
+            }
+          }
+
+          loadedPids.foreach {
+            case (transactionalId, txnMetadata) =>
+              val currentTxnMetadata = addTransaction(transactionalId, txnMetadata)
+              if (!txnMetadata.equals(currentTxnMetadata)) {
+                // treat this as a fatal failure as this should never happen
+                fatal(s"Attempt to load $transactionalId's metadata $txnMetadata failed " +
+                  s"because there is already a different cached pid metadata $currentTxnMetadata.")
+
+                throw new KafkaException("Loading transaction topic partition failed.")
+              }
+          }
+
+          removedPids.foreach { transactionalId =>
+            // if the cache already contains a pid which should be removed, raise an error.
+            if (transactionMetadataCache.contains(transactionalId))
+              throw new IllegalStateException(s"Unexpected unload of $transactionalId's pid metadata while " +
+                s"loading partition $topicPartition")
+          }
+
+          info(s"Finished loading ${loadedPids.size} pid metadata from $topicPartition in ${time.milliseconds() - startMs} milliseconds")
+        }
+    }
+  }
+
+  /**
+    * When this broker becomes a leader for a transaction log partition, load this partition and
+    * populate the pid metadata cache with the transactional ids.
+    */
+  def loadTransactionsForPartition(partition: Int) {
+    validateTransactionTopicPartitionCountIsStable()
+
+    val topicPartition = new TopicPartition(Topic.TransactionStateTopicName, partition)
+
+    def loadPidAndTransactions() {
+      info(s"Loading pid metadata from $topicPartition")
+
+      inLock(partitionLock) {
+        if (loadingPartitions.contains(partition)) {
+          // with background scheduler containing one thread, this should never happen,
+          // but just in case we change it in the future.
+          info(s"Pid and transaction status loading from $topicPartition already in progress.")
+          return
+        } else {
+          loadingPartitions.add(partition)
+        }
+      }
+
+      try {
+        loadTransactionMetadata(topicPartition)
+      } catch {
+        case t: Throwable => error(s"Error loading offsets from $topicPartition", t)
```

Comment:
Wrong error message: should be "Error loading transactions from transaction log $topicPartition"